03_Modules/SmartCharging/src/module/api.ts

```diff
@@ -91,12 +97,372 @@ export class SmartChargingModuleApi
     CallAction.SetChargingProfile,
     SetChargingProfileRequestSchema,
   )
-  setChargingProfile(
+  async setChargingProfile(
     identifier: string,
     tenantId: string,
     request: SetChargingProfileRequest,
     callbackUrl?: string,
   ): Promise<IMessageConfirmation> {
+    this._logger.info(
+      `Received SetChargingProfile: ${JSON.stringify(request)}`,
+    );
+
+    const chargingProfile: ChargingProfileType = request.chargingProfile;
+
+    // Validate Charging Profile Stack Level
+    if (chargingProfile.stackLevel < 0) {
+      return {
+        success: false,
+        payload: 'Stack Level SHALL be >= 0.',
+      };
+    }
+
+    // Validate Charging Profile valid period
+    const now = new Date();
+    const validFrom = chargingProfile.validFrom
+      ? new Date(chargingProfile.validFrom)
+      : null;
+    const validTo = chargingProfile.validTo
+      ? new Date(chargingProfile.validTo)
+      : null;
+    // OCPP 2.0.1 Part 2 K01.FR.36
+    if (validFrom && validFrom.getTime() > now.getTime()) {
+      return {
+        success: false,
+        payload: `chargingProfile validFrom ${chargingProfile.validFrom} should not be in the future.`,
+      };
+    }
+    // OCPP 2.0.1 Part 2 K01.FR.37
+    if (validTo && validTo.getTime() <= now.getTime()) {
+      return {
+        success: false,
+        payload: `chargingProfile validTo ${chargingProfile.validTo} should be in the future.`,
+      };
+    }
+
+    // Validate Charging Profile Purpose
+    let receivedChargingNeeds;
+    let transactionDatabaseId;
+    if (
+      chargingProfile.chargingProfilePurpose ===
+      ChargingProfilePurposeEnumType.TxProfile
+    ) {
+      if (!chargingProfile.transactionId) {
+        return {
+          success: false,
+          payload:
+            'Missing transactionId for chargingProfilePurpose TxProfile.',
+        };
+      }
+      if (
+        !(await this._module.transactionEventRepository.readTransactionByStationIdAndTransactionId(
+          identifier,
+          chargingProfile.transactionId,
+        ))
+      ) {
+        return {
+          success: false,
+          payload: `Transaction ${chargingProfile.transactionId} not found on station ${identifier}.`,
+        };
+      }
+      if (request.evseId <= 0) {
```

Comment:
Nitpick. Quick comment of requirement reference
```suggestion
      // OCPP 2.0.1 Part 2 K01.FR.16
      if (request.evseId <= 0) {
```